src/main/java/com/google/api/gax/grpc/InstantiatingChannelProvider.java

```diff
@@ -0,0 +1,308 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.grpc;
+
+import com.google.api.gax.core.CredentialsProvider;
+import com.google.auth.Credentials;
+import com.google.common.collect.Lists;
+import io.grpc.ClientInterceptor;
+import io.grpc.ManagedChannel;
+import io.grpc.auth.ClientAuthInterceptor;
+import io.grpc.netty.NegotiationType;
+import io.grpc.netty.NettyChannelBuilder;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.Executor;
+
+public class InstantiatingChannelProvider implements ChannelProvider {
+  private static final String DEFAULT_GAX_VERSION = "0.1.0";
+
+  private final ExecutorProvider executorProvider;
+  private final CredentialsProvider credentialsProvider;
+  private final String serviceAddress;
+  private final int port;
+  private final String clientLibName;
+  private final String clientLibVersion;
+  private final String serviceGeneratorName;
+  private final String serviceGeneratorVersion;
+
+  private InstantiatingChannelProvider(
+      ExecutorProvider executorProvider,
+      CredentialsProvider credentialsProvider,
+      String serviceAddress,
+      int port,
+      String clientLibName,
+      String clientLibVersion,
+      String serviceGeneratorName,
+      String serviceGeneratorVersion) {
+    this.executorProvider = executorProvider;
+    this.credentialsProvider = credentialsProvider;
+    this.serviceAddress = serviceAddress;
+    this.port = port;
+    this.clientLibName = clientLibName;
+    this.clientLibVersion = clientLibVersion;
+    this.serviceGeneratorName = serviceGeneratorName;
+    this.serviceGeneratorVersion = serviceGeneratorVersion;
+  }
+
+  @Override
+  public boolean needsExecutor() {
+    return executorProvider == null;
+  }
+
+  @Override
+  public ManagedChannel getChannel() throws IOException {
+    if (needsExecutor()) {
+      throw new IllegalStateException("getChannel() called when needsExecutor() is true");
+    } else {
+      return createChannel(executorProvider.getExecutor());
+    }
+  }
+
+  @Override
+  public ManagedChannel getChannel(Executor executor) throws IOException {
+    if (!needsExecutor()) {
+      throw new IllegalStateException("getChannel(Executor) called when needsExecutor() is false");
```

Comment:
Shall we throw an exception or we should allow user to override the executor here